posts/Error-Handling-Patterns-in-Expressjs.md

```diff
@@ -0,0 +1,234 @@
+# Error Handling Patterns in Express.js
+
+
+Building an application isn't just about writing code that works. It's about making your program run flawlessly, so it behaves well when you encounter unexpected conditions. There are at large two types of errors in JavaScript: programmer errors and real-time errors. The programmer errors are fixable during development, while real-time, operational errors are not predictable. You need an efficient error handling mechanism that keeps your Node server stable regardless of its type.
+
+The objective of this tutorial is to teach you different ways that you can handle JavaScript errors for your Express server. 
+
+## Preqrequisites
+
+You need to set up an Express server to get started. Assuming that you've already installed Node.js and npm, create a working directory for your application. 
+
+```bash
+mkdir demo-app
+cd demo-app
+```
+
+Use the `npm install` script to generate a *package.json* file. I prefer to set the entry point of the application to *app.js*. Next, install express and other related packages and save these dependencies into your *package.json*.
+
+```bash
+npm install express --save
+npm install -g nodemon
+```
+I've installed nodemon to monitor the changes in the demo-app project and auto restart the node.js application. 
+
+Here is a very basic middleware to test that everything's in place:
+
+```
+const express = require('express')
+const app = express()
+
+app.get('/', (req, res) => res.json(
+    {
+        status: 200,
+        message: "Server is up and running"
+    })
+
+app.listen(3000, () => console.log('Listening to port 3000!'))
+```
+
+## A bit of theory first
+
+Let's have a look at what happens under the hood when you run your application. When an error is encountered in JavaScript, the interpreter stops whatever it's doing and then throws an exception. The throw here refers to JavaScript's throw mechanism. The execution of the current function stops and the interpreter starts looking for a catch block so that the application doesn't terminate. However, when it doesn't find one, it jumps out of the current function and all the other functions on top of it in the call stack until an exception handler is found. When it reaches the top-level function that started the execution, it terminates the program because no exception handling mechanism was found. That's when Node returns an error screen in the console.
```

Comment:
I think this would be a great place for a picture. You can make them using LucidChart for free.